scripts-base/src/Component.ts

```diff
@@ -0,0 +1,96 @@
+import {
+	ComponentEl,
+	DelegateEvent,
+	DelegateTarget,
+	EventListeners,
+	EventMap,
+	EventMapByElement,
+} from './componentTypes'
+import { matchesSelector } from './utils'
+
+export interface NamedComponent {
+	componentName: string
+}
+
+export type ComponentConstructor<
+	D,
+	E extends ComponentEl,
+	EMap extends EventMap = EventMapByElement<E>
+> = NamedComponent & (new (element: E, data: D) => Component<D, E, EMap>)
+
+export class ComponentInitializationError extends Error {}
+
+export class Component<D = {}, E extends ComponentEl = HTMLElement, EMap extends EventMap = EventMapByElement<E>> {
+	protected readonly getListeners = (): EventListeners<E, EMap> => []
+
+	constructor(protected readonly el: E, protected readonly data: D) {}
+
+	public setup() {
+		this.attachListeners()
+
+		this.init()
+	}
+
+	protected getChild<C extends DelegateTarget<E>>(
+		selector: string,
+		ChildConstructor: new (...args: any[]) => C,
```

Comment:
Co je to `ChildConstructor`? K čemu to mám používat?

Reply:
Tohle je určeno pro případy, kdy komponenta *nezbytně* potřebuje referenci na child element, např. kvůli tomu, že bez toho ta komponenta ani nemůže fungovat. To schodou okolností nastává takřka vždy, když tu referenci potřebuje, tedy že drtivá většina referencí na children jsou nezbytné. Když člověk volá `this.el.querySelector('...')`, vrátí to např. `HTMLDivElement` nebo `null`. U nezbytných referencí by člověk musel psát něco jako:
```ts
private childDiv: HTMLDivElement
private childSpan: HTMLSpanElement

public constructor(el, data) {
	const childDiv = this.el.querySelector('…')
	const childSpan = this.el.querySelector('…')

	if (!(childDiv instanceof HTMLDivElement) || !(childSpan instanceof HTMLSpanElement)) {
		throw new ComponentInitializationError()
	}

	this.childDiv = childDiv
	this.childSpan = childSpan
}
```
Místo toho může napsat
```ts
private childDiv: HTMLDivElement
private childSpan: HTMLSpanElement

public constructor(el, data) {
	this.childDiv = this.getChild('…', HTMLDivElement)
	this.childSpan = this.getChild('…', HTMLSpanElement)
}
```
a neřešit zbytečné ifování.

Reply:
Ta podmínka, že se `this.getChild` může používat jen v konstruktoru mi přijde zbytečně omezující. Proč tahle limitace vůbec vznikla?

Co když chci získat referenci na element, který v době volání konstruktoru neexistuje? Třeba u nějakého ajaxového obsahu, se kterým se pracuje až po jeho načtení. Nebo mám tlačítko, na které se často nekliká a dělá nějakou jednorázovou akci nad potomky, takže je zbytečné zpomalovat konstruktor, je lepší náročnou akci dělat jen v případě potřeby.